### PR TITLE
feat(jax): add graph_mode kwarg to neighbor lists for CUDA graph replay

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Adds an opt-in `graph_mode={"none","warp"}` kwarg to the JAX naive and cell-list neighbor-list APIs so the underlying Warp kernels can be fused into a single CUDA-graph capture+replay via `jax_callable(..., graph_mode=GraphMode.WARP)` instead of N separate `jax_kernel` launches. To make the wrapped-PBC naive path graph-cacheable, `naive_neighbor_list` also accepts pre-allocated `inv_cell`, `positions_wrapped`, and `per_atom_cell_offsets` buffers (closure-captured by the caller so the buffer pointers stay stable across calls). Also deprecates the legacy `zero_array` helper in favor of Warp's built-in `wp.array.zero_()`.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

N/A

## Changes Made

- Added `graph_mode: Literal["none", "warp"] = "none"` to `naive_neighbor_list`, `build_cell_list`, `query_cell_list`, and `cell_list` in `nvalchemiops/jax/neighbors/{naive,cell_list}.py`. The `"warp"` path registers a fused `jax_callable(..., graph_mode=GraphMode.WARP)` per (dtype, selective-rebuild, return-mode) variant so steady-state calls become a CUDA-graph replay.
- Extended `naive_neighbor_list` with optional `inv_cell`, `positions_wrapped`, and `per_atom_cell_offsets` kwargs so callers can pre-allocate the wrap-kernel scratch buffers and keep their pointers stable across `graph_mode="warp"` replays. The new docstring example shows the closure-capture + `donate_argnums` pattern.
- Deprecated `nvalchemiops.neighbors.neighbor_utils.zero_array`: it now emits a `DeprecationWarning` pointing at `wp.array.zero_()`, and the symbol was dropped from the `nvalchemiops.neighbors.__init__` and `nvalchemiops.neighborlist.__init__` re-exports so new code is steered to the built-in. The function itself is still importable from `nvalchemiops.neighbors.neighbor_utils` for backwards compatibility. Internal call sites in `nvalchemiops/neighbors/{cell_list,batch_cell_list}.py` and `benchmarks/dynamics/shared_utils.py` switched to `array.zero_()`.
- Added `TestNaiveGraphMode` (22 parameterized cases incl. parity with the default path, `return_neighbor_list=True`, the donation/closure contract via `test_wrapped_warp_replay_stable_pointers`, and `ValueError` on invalid values) to `test/neighbors/bindings/jax/test_naive.py`.
- Added `TestCellListGraphMode` (13 cases across `build_cell_list` / `query_cell_list` / `cell_list` parity and invalid-value handling) to `test/neighbors/bindings/jax/test_cell_list.py`.
- Removed the no-longer-needed `test_zero_array` / `test_zero_array_empty` cases from `test/neighbors/test_neighbor_utils_kernels.py`.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Latency benchmark on a single GPU with 200 atoms (median single-call wall-clock, ms; lower is better; JAX 0.9.0.1, Warp 1.11.0, RTX A6000 Ada). The "Warp-kernel share" column is the pure CUDA-kernel time (CUDA-event floor) divided by the measured wall, i.e. the fraction of each call that is actually GPU work; the remainder is host-side dispatch overhead.

| Public API path                     | `graph_mode="none"` | `graph_mode="warp"` (replay) | Speedup | Warp-kernel share (none → warp) |
| ----------------------------------- | ------------------: | ---------------------------: | ------: | ------------------------------: |
| `naive_neighbor_list` (no PBC wrap) | 0.160 ms            | 0.139 ms                     | 1.15x   | 59% → 80%                       |
| `naive_neighbor_list` (PBC wrap)    | 0.234 ms            | 0.143 ms                     | 1.64x   | 47% → 84%                       |
| `build_cell_list`                   | 0.158 ms            | 0.063 ms                     | 2.52x   | 34% → 87%                       |
| `query_cell_list`                   | 0.213 ms            | 0.192 ms                     | 1.11x   | 64% → 71%                       |

Reference (lower bound): the hand-fused prototype `jax_callable_wrap_and_fill_graphwarp_replay` reaches 0.131 ms on the wrapped-naive path; the public `naive_neighbor_list(..., graph_mode="warp", wrap_positions=True)` lands at 0.143 ms (~9% off the prototype). The rising Warp-kernel share confirms the host-dispatch overhead is what's being eliminated rather than any GPU-side speedup. The wrapped-PBC replay required closure-capturing the `inv_cell` / `positions_wrapped` / `per_atom_cell_offsets` scratch buffers so their pointers stay stable across calls; without that fix, the Warp graph cache misses every step and replay regresses to ~0.185 ms.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

`graph_mode` defaults to `"none"`, so this is a purely opt-in addition with no behavior change for existing callers. `zero_array` is still importable from `nvalchemiops.neighbors.neighbor_utils` (now emits a `DeprecationWarning`) so existing user code keeps working; downstream callers should migrate to the Warp built-in `wp.array.zero_()` (one-liner: `my_array.zero_()`, no `device` arg needed since the array already knows its device, and faster because there's no Python-side kernel dispatch). The closure-capture requirement for `inv_cell` / `positions_wrapped` / `per_atom_cell_offsets` is explicitly documented in `naive_neighbor_list`'s docstring under "Notes" and demonstrated in the docstring example, mirroring the `_make_public_naive_stateful` helper in the new latency benchmark.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/jax/neighbors/cell_list.py [hunk 1]
```diff
@@ -116,6 +127,528 @@
...
+        cutoff,
+        cells_per_dimension,
+        atom_periodic_shifts,
+        atom_to_cell_mapping,
+        atoms_per_cell_count,
+        cell_atom_start_indices,
+        cell_atom_list,
+        wp_dtype,
+        str(positions.device),
+    )
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/cell_list.py:177]: Fixed in fa12836cbc47caa84688201d39b45f002298ddf5.

File: nvalchemiops/jax/neighbors/naive.py [hunk 2]
```diff
@@ -399,8 +1095,129 @@ def naive_neighbor_list(
...
+                    half_fill,
+                )
+        else:
+            if cell.dtype != positions.dtype:
+                cell = cell.astype(positions.dtype)
+
+            num_shifts = int(max_shifts_per_system)
+            if wrap_positions:
+                if inv_cell is None:
+                    inv_cell = jnp.linalg.inv(cell)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/naive.py:1186]: Fixed in fa12836cbc47caa84688201d39b45f002298ddf5.

File: nvalchemiops/jax/neighbors/naive.py [hunk 3]
```diff
@@ -144,6 +147,616 @@
 )
 
 
+def _validate_graph_mode(graph_mode: str) -> Literal["none", "warp"]:
+    """Validate the public graph-mode string."""
+    if graph_mode not in {"none", "warp"}:
+        raise ValueError("graph_mode must be one of {'none', 'warp'}")
+    return graph_mode
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/naive.py]: Fixed in fa12836cbc47caa84688201d39b45f002298ddf5.

**Comment by dallasfoster** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/commit/fa12836cbc47caa84688201d39b45f002298ddf5

File: nvalchemiops/jax/neighbors/naive.py [hunk 4]
```diff
@@ -290,12 +930,64 @@ def naive_neighbor_list(
...
+    ...         cell=cell,
+    ...         pbc=pbc,
+    ...         neighbor_matrix=neighbor_matrix,
+    ...         num_neighbors=num_neighbors,
+    ...         neighbor_matrix_shifts=shifts,
+    ...         inv_cell=inv_cell,
+    ...         positions_wrapped=positions_wrapped,
+    ...         per_atom_cell_offsets=per_atom_cell_offsets,
+    ...         graph_mode="warp",
+    ...     )
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/naive.py:965]: Fixed in deea0899d5717fc3a83dbdf4ac9c0a3ab64f44a4.

**Review by dallasfoster — APPROVED** [review]
